File: arm.64/frame/blazor/src/type.cs
using System;
using System.Runtime.InteropServices;

namespace Pat {
    public delegate bool FitFn(IntPtr address, IntPtr payload, IntPtr options);
    public delegate int FunFn(IntPtr address, IntPtr payload, IntPtr options);

    [StructLayout(LayoutKind.Sequential)]
    public struct Def {
        public IntPtr sid;
        public IntPtr tag;
        public IntPtr fit;
        public IntPtr fun;
    }

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate int MoreFn(Def def);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate int PumpFn(
        [MarshalAs(UnmanagedType.LPUTF8Str)] string address,
        [MarshalAs(UnmanagedType.LPUTF8Str)] string payload,
        [MarshalAs(UnmanagedType.LPUTF8Str)] string options);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate int LessFn(Def def);

    public class Cabi {
        public string Name { get; set; } = "";
        public string Path { get; set; } = "";
        public IntPtr Lib { get; set; }
        public MoreFn? More { get; set; }
        public PumpFn? Pump { get; set; }
        public LessFn? Less { get; set; }
    }

    public class Triplet {
        public string Address { get; set; }
        public string Payload { get; set; }
        public string Options { get; set; }

        public Triplet(string address, string payload = "{}", string options = "{\"once\":true}") {
            Address = address;
            Payload = payload;
            Options = options;
        }
    }
}

File: arm.64/frame/avalonia/src/type.fs
namespace Pat

open System
open System.Runtime.InteropServices

type FitFn = delegate of address: nativeint * payload: nativeint * options: nativeint -> bool
type FunFn = delegate of address: nativeint * payload: nativeint * options: nativeint -> int

[<Struct; StructLayout(LayoutKind.Sequential)>]
type Def = {
    sid: nativeint
    tag: nativeint
    fit: nativeint
    fun_: nativeint
}

[<UnmanagedFunctionPointer(CallingConvention.Cdecl)>]
type MoreFn = delegate of def_: Def -> int

[<UnmanagedFunctionPointer(CallingConvention.Cdecl)>]
type PumpFn = delegate of [<MarshalAs(UnmanagedType.LPUTF8Str)>] address: string * [<MarshalAs(UnmanagedType.LPUTF8Str)>] payload: string * [<MarshalAs(UnmanagedType.LPUTF8Str)>] options: string -> int

[<UnmanagedFunctionPointer(CallingConvention.Cdecl)>]
type LessFn = delegate of def_: Def -> int

type Cabi = {
    Name: string
    Path: string
    Lib: nativeint
    More: MoreFn
    Pump: PumpFn
    Less: LessFn
}

type Triplet = {
    Address: string
    Payload: string
    Options: string
}
